### PR TITLE
Ralph2 vs Ralph3 cross validator

### PR DESCRIPTION
**Pull request by mkurek**
MVP of data validator between ralph2 and ralph3. Currently it supports AssetModel and DataCenterAsset (partialy).

### CODE REVIEW

File: src/ralph/cross_validator/__main__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import os
+import sys
+
+
+def main():
+    if 'cross_validator' not in os.environ.get('DJANGO_SETTINGS_MODULE', ''):
```
Comment (ghost) [inline src/ralph/cross_validator/__main__.py:6]: is this correct?

Comment (mkurek) [inline src/ralph/cross_validator/__main__.py:6]: yes - it is correct (but not perfect). the idea here is to enable local settings easily (ex. ralph.cross_validator.settings_local).

File: src/ralph/cross_validator/db_routers.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class CrossRouter(object):
+    def db_for_read(self, model, **hints):
+        if hasattr(model, 'i_am_your_father'):
```
Comment (szok) [inline src/ralph/cross_validator/db_routers.py:3]: +1

File: src/ralph/cross_validator/helpers.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    common_keys = set(old.keys()).intersection(set(new.keys()))
+    common_keys -= set(blacklist) if blacklist else set()
+    for change in common_keys:
+        old_value = normalize(old[change])
+        new_value = normalize(new[change])
+        if old_value != new_value:
+            yield change, old_value, new_value
+
+
+def getattr_dunder(obj, field_path, default=None):
```
Comment (szok) [inline src/ralph/cross_validator/helpers.py]: getattr_dunder already exists in: https://github.com/allegro/ralph/blob/d3cdc632fd4cf5f2416c29736f272d85adfed988/src/ralph/admin/helpers.py#L58

File: src/ralph/cross_validator/ralph2/device.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    class Meta(generate_meta(app_label='ralph_assets', model_name='rack')):
+        pass
+
+
+class DeviceInfo(SoftDeletable, models.Model):
+    ralph_device = models.OneToOneField(Device)
+    u_level = models.CharField(max_length=10, null=True, blank=True)
+    u_height = models.CharField(max_length=10, null=True, blank=True)
+    # data_center = models.ForeignKey(DataCenter, null=True, blank=False)
+    # server_room = models.ForeignKey(ServerRoom, null=True, blank=False)
```
Comment (szok) [inline src/ralph/cross_validator/ralph2/device.py]: remove comments?

File: src/ralph/cross_validator/admin.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        return super().get_queryset(*args, **kwargs).select_related(
+            'content_type'
+        )
+
+    def get_diff_display(self, obj):
+        if not bool(obj.diff) or bool(obj.errors):
+            return format_html('-')
+        html = ''
+        for item, values in obj.diff.items():
+            print(item, values)
```
Comment (szok) [inline src/ralph/cross_validator/admin.py]: Remove print

File: src/ralph/cross_validator/admin.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                mono_end='</span>',
+            )
+        return format_html(html)
+    get_diff_display.short_description = 'Diff'
+
+    def get_errors_display(self, obj):
+        if not bool(obj.errors):
+            return '-'
+        html = ''
+        for item in obj.errors:
```
Comment (szok) [inline src/ralph/cross_validator/admin.py]: Maybe
'''
html = "<br>".join(obj.errors)
'''

File: src/ralph/cross_validator/admin.py (hunk 5 above)
Comment (ghost) [inline src/ralph/cross_validator/admin.py]: http://flake8.pycqa.org/en/latest/extensions.html#existing-extensions
maybe we should use anh of them, especially flake8-print?

**Comment by ghost** [comment]
:ok:

File: src/ralph/cross_validator/admin.py (hunk 5 above)
Comment (mkurek) [inline src/ralph/cross_validator/admin.py]: 👍

**Comment by szok** [comment]
:ok:

File: src/ralph/cross_validator/admin.py (hunk 5 above)
Comment (mkurek) [inline src/ralph/cross_validator/admin.py]: @xliiv https://github.com/allegro/ralph/pull/2492

**Pull request merged**